---
title: "Family Project Tutorial"
author: '[authors]'
date: "TBD"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The primary goal of this addendum is to foster familiarity with basic microbial data analysis and figure generation. The data provided to perform this exercise will be used to demonstrate how to create figures in R similar to those found in the main publication.**[more explicit]** In theory, the concepts found in this guide should be applicable to other microbial datasets with some degree of modification.  In order to complete this tutorial basic knowledge of R and its functions is assumed. In this guide concepts are explained at an introductory level supplemented by sample R code for the provided activity. Statistical explanations found in this guide are by no means comprehensive and it is the responsibility of the reader to ensure that the assumptions of any test performed are appropriate for the given data. 

## R Nomenclature 

Here are some tips to help get through RStudio if it’s your first time using it.

1. To see more information on a *function*, such as the inputs taken or output, use:

`help(function)`
or
`?function`

2. Run a single line by selecting it and pressing CTRL + ENTER
3. Regions of code that are followed by a # are comments and are not necessary for the code to run. They are to explain what the code is doing e.g.: 
```{r}
df <- mtcars$mpg #assign the mpg column of dataframe mtcars to df
```
4. To complement understanding of particular functions, whenever an existing data frame is modified or a new one is created visualize it either by clicking the data frame in the environment or using the View function e.g.: `View(df)`

## Preparing Data for Figures 

 To create figures it is important to ensure that the data is prepared and formatted correctly. Before beginning, specify a desired working directory in R-studio. The metadata, taxonomy, and OTU tables found at **(location we are making them accessible)** are the only files needed for the completion of this tutorial. 
 
### Data Import 

First set the appropriate working directory, i.e. where we want R to look for our files. Change `File Path` to the location where your desired data is being kept. *It's my file path for now but when were done change back to "File Path".*
```{r}
setwd("C:/Users/Zach/Documents/Microbiome Family Project")
```

Import the metadata file, taxonomy, and OTU Table. For the taxonomy and OTU table the first column can be specified as the row names, the metadata file will require some extra steps to be formatted properly. 
```{r}
library(biomformat)

OTU.taxonomyFP <- read.delim("taxonomy.txt", row.names = 1)
my_biom <- read_hdf5_biom("61277_otu_table.biom") #import biom
write_biom(my_biom, "formatted_biom.biom") #format biom
my_biom <- read_biom("formatted_biom.biom") #read formatted biom
OTU.tableFP <- as.data.frame(as.matrix(biom_data(my_biom))) #create OTU table from biom
colnames(OTU.tableFP) <- sub('11959.', 'X', colnames(OTU.tableFP)) #reformat columns
metadataFP <- read.delim("Microbiome_mapping_14Oct2019.txt")
```

Ensure that all the desired files maintain a consistent formatting style. For the metadata data frame the row names must be formatted for consistency across the other files. The row names must be in the format of “X.##” -  “##” being indicative of the sample number. 
```{r}
metadataFP$NAME <- interaction("X", metadataFP$NAME, sep = "")
rownames(metadataFP) <- metadataFP[,1]
metadataFP[,1] <- NULL
```

### Data Filtering 

It is important to determine if filtering is needed before continuing. An example utilizing the sample data is provided below, however, note that this procedure will not be implemented when creating the final figures at later stages in this tutorial. 

Merge the OTU table with taxonomy into a single data frame. 
```{r}
OTU.table.with.taxonomyFP <- as.data.frame(merge(OTU.taxonomyFP, 
                                               OTU.tableFP, by.x = "row.names", by.y = "row.names"))
```

Remove unassigned taxa from the merged data frame. For certain diversity metrics filtering may be undesired **citation needed**
```{r}
OTU.table.filteredtaxaFP <- OTU.table.with.taxonomyFP[!grepl("Unassigned", 
                                                        OTU.table.with.taxonomyFP$Kingdom),]
```

Subset the original OTU table with the filtered taxa to create a filtered OTU table. The filtered OTU table would be used in place of the OTU table in the subsequent steps to generate a rarefied OTU table, given that filtration is desired. 
```{r}
FPfilteredtaxa <- OTU.table.filteredtaxaFP$Row.names
FPfilterOTU <- subset(OTU.tableFP, rownames(OTU.tableFP) %in% FPfilteredtaxa)
```

### Rarefaction 
A re-sampling technique for richness estimation, term also used to reference a normalization technique. **go into more depth with the explanation later**

Disclaimer for now: This section may be subject to significant change. 

First retrieve the quartile ranges to rarefy by. Note that the ranges depend on the depth of the samples to be kept. In this example, `Q10` and `Q15` discard the bottom 10th and 15th percentiles respectively. Use the appropriate ranges for your data. 

```{r}
transOTU <- rowSums(t(OTU.tableFP)) 
Q10 <- quantile(transOTU[order(transOTU, decreasing = TRUE)], 0.10)
Q15 <- quantile(transOTU[order(transOTU, decreasing = TRUE)], 0.15) 
```

Visualize these ranges with a barplot of our samples
```{r}
barplot(sort(transOTU), ylim = c(0, max(transOTU)), 
        xlim = c(0, NROW(transOTU)), col = "Blue", ylab = "Read Depth", xlab = "Sample") 
abline(h = c(Q10, Q15), col = c("red", "pink"))
```

Generate a rarefaction curve. If the samples begin to plateau, this is a good indication that the samples were sequenced adequately, as it is indicative of more accurately capturing the diversity of a community. 
```{r, message=FALSE}
library(vegan)
```
```{r}
rarecurve(t(OTU.tableFP), step = 100, cex = 0.5)
abline(v = c(Q10, Q15), col = c("red", "pink"))
```
  
As you can see, these two lines serve as potential cutoff points for the samples to the left of each line. It is a good idea to visualize the samples with a rarefaction curve like this and make a decision about where to cut off samples. After taking a closer look, a more optimal position to cut off samples would be at a sample size of ~1350. Here you can see that only the 1 sample with insufficient sample depth is not included in the final rarefied data. Here's what rarefying to ~1350 reads per sample size would look like.
```{r}
transOTU <- rowSums(t(OTU.tableFP)) 
Q01 <- quantile(transOTU[order(transOTU, decreasing = TRUE)], 0.0106) #~1350 reads per sample 
#^ is this quantile notation in the beginning correct
rarecurve(t(OTU.tableFP), step = 100, cex = 0.5)
abline(v = c(Q01), col = c("purple"))
```

The next step involves cutting off the rarer data. Here we want to make as best a decision we can to keep as much data as possible while rarefying the data.

```{r}
library(EcolUtils)
```
```{r, message=FALSE, warning=FALSE, results=FALSE}
FPrared.OTU <- as.data.frame((rrarefy.perm(t(OTU.tableFP), sample = Q01, n = 100, round.out = T))) #with all samples - rarefies the data

FPrared.OTU <- as.data.frame(FPrared.OTU[rowSums(FPrared.OTU) >= Q01 - (Q01 * 0.0106), colSums(FPrared.OTU) >= 1]) #with some samples cut off - removes low depth samples that do not meet the cutoff
```
## Alpha Diversity

Alpha diversity refers to microbial diversity found within a sample or in the environment. Each alpha diversity sample contains richness metrics which quantify the OTUs, or unique species present within a sample, and evenness metrics which evaluate how equally each species is represented in a sample. Three commonly used indices: Shannon, Simpson, and Chao may produce different results based on the weights placed on evaluation metrics. Of these indices, Chao is incapable of capturing the scope of microbial diversity and as such the Shannon and Simpson indices will be further explored (Qian et al., 2020).

Shannon and Simpson diversity both calculate alpha diversity through a combination of richness and evenness metrics. However, they differ in regard to the weight assigned to certain species types. The Shannon index ranges from 0-5, with higher values indicative of greater diversity, however greater weight is placed on presence of rarer taxa which would increase this value. In contrast to this the Simpson index ranges from 0-1 but emphasizes common taxa instead (Citation). 

It is best to consider which particular index applies to each dataset and utilize whichever produces the most sound results. In this study the Shannon index was utilized to generate an alpha diversity figure between the four familial cohorts. 

### Generating the α-diversity plot in R 

Create a richness table from the rarefied OTU table. Not that rarefaction may induce changes in diversity metrics, ensure that it is done properly for the dataset of interest. 
Rename the column to something identifiable, in this case “speciesrich” 
```{r}
FPrichness <- as.data.frame(specnumber(FPrared.OTU))
colnames(FPrichness) <- c("speciesrich")
FPshannon <- as.data.frame(diversity(FPrared.OTU, index = "shannon"))
colnames(FPshannon) <- c('alpha_shannon')
```

## Beta Diversity

Beta diversity shows the difference in diversity between two samples or environments. Two common forms of beta diversity are PCoA and NMDS. With each of these visuals, one can get a sense of how similar samples are by how close in proximity points appear on these graphs. There are several different ways to compose beta diversity plots, which vary in both dimensions used and the distance metric applied. With NMDS in particular, two environments are compared by a 
certain number of dimensions, usually narrowed down as much as possible for a correlation to be made, but also still retaining the 
accuracy of the initial data. Each beta diversity plot utilizes a distance 
metric, for example, Euclidean distance for PCoA. 
#Ordination Methods

As mentioned above, there are various beta diversity methods, the most popular of which are PCoA and NMDS. In the tutorial below, a distance metric called Bray-Curtis dissimilarity is used. 

The beta diversity tutorial will use the Family Project rarefied OTU table (produce this in rarefaction section first) and the microbiome mapping data.

Initialize plot with rarefaction OTU table, bray-curtis dissimilarity and 2 dimensions(k):
```{r, results=FALSE}
NMDS1 <- metaMDS(FPrared.OTU, distance = "bray", k = 2)
```

Extract the two axes from the NMDS to plot x and y coordinates
```{r}
coordinates <- data.frame(NMDS1$points[,1:2])
```


Plot - MDS1 is the x axis, MDS2 is the y axis. This is not the complete plot however, microbiome mapping must be added next.
```{r}
plot(x = coordinates$MDS1, y = coordinates$MDS2)
```

Merge the NMDS with Microbiome_mapping:
```{r}
nmds_plus_metadataFP <- merge(coordinates, metadataFP, by = 0)
```

Choose a factor to compare the environments. In this case, we compare them by “Family”:
```{r}
Factor_x <- as.factor(nmds_plus_metadataFP$Family)

filtersamples <- rownames(FPrared.OTU)
filtermapping <- subset(metadataFP, rownames(metadataFP) %in% filtersamples)
```

Adonis test (explain more):
```{r, results=FALSE}
adonis(data = filtermapping, formula = FPrared.OTU ~ Family 
       / Individual + Day + Day:Family,
       permutations = 999, method = "bray")

adonis(data = filtermapping, formula = FPrared.OTU ~ Family
       /Individual/Day,
       permutations = 999, method = "bray")
```


Plot with rarefaction:
```{r}
ggplot(data = nmds_plus_metadataFP) +
  aes(x = MDS1, y = MDS2, color = Family) +
  geom_point() +
  labs(fill = "Family") +
  ggtitle("NMDS of Families on Oral Microbiome") + #adds title
  theme_classic(base_size = 14, base_line_size = .5) +
  stat_ellipse(size = 1, type = "t", level = .95)
```


## Random Forest

Random forest is a machine learning technique that in this instance is used to determine the most influential taxa in distinguishing between families. The algorithm classifies samples into groups based on similar features though the use of decision trees (Zhou & Gallins, 2019).
When it comes to evaluating data based on a decision tree, not all data can be accurately evaluated. That's where random forests come in as a collection of hundreds of decision trees that each data sample is run through. When each data sample is run through the entire random forest, an aggregate decision is made for that sample. This is known as "Bagging". The aggregate decision in this random forest would be which family that sample belongs to. After that is all said and done, we can calculate how accurate our random forest is by checking the out-of-box error. 
Two packages are needed
to perform a random forest. Ensure that the `randomForest` and `rfPermute`
packages are installed, then load them for use.

```{r, message=FALSE,warning=FALSE, results=FALSE}
library(randomForest)
library(rfPermute)
```

Import the Family Phyla file and format it so that the family column is a factor. *we would need to put the all the files, including the ones for this random forest section in one directory*

```{r}
family_Phyla <- read.csv(file="C:/Users/Zach/Documents/Microbiome Family Project/rfPermute_Tutorial/family_Phyla.csv",
header =T, sep=",")
family_Phyla$Family <- factor(family_Phyla$Family)
```
Create the random forest specifying the factor that it should be clustered by - in this case Family.
```{r}
family_rfp <-
rfPermute(Family~., data = family_Phyla)
```
Next visualize the variable importance plot. This will tell us which Phyla is most important in determining the family that each sample belongs to.
```{r}
#pdf("Variable_importance_generated.pdf", width = 5, height = 5) # remove the "#" in front if this is to be generated as a separate pdf 
varImpPlot(family_rfp, type=1)
print(family_rfp)
```

Create and visualize the proximity plot. This shows us the similarity of samples with respect to each other and are colored based on their respective families. Samples that are closer are more similar.
```{r}
family_rfp1 <-
rfPermute(Family~., data = family_Phyla, proximity = T)
#pdf("proximity_plot_generated.pdf", width=5, height=5) # remove the "#" in front if this is to be generated as a separate pdf
proximityPlot(family_rfp1)
print(family_rfp1)
```

## Creating an Abundance Bar-plot -->

<!-- In this section we are going to go over the creation of bar-plot to demonstrate the distribution of the most prevalent taxa in each section. In order to do this two methods will be demonstrated, a quick an convenient method utilizing phyloseq, and a more detailed process using a host of packages. For some users the plot generated by phyloseq may be all that neccesary, however, the second method using the tidyverse, reshape2 and zoo packages allows for a more granular level of control. The reshape2 package is particularly important as it allows us generate a melted (long format) table of all our data. The zoo package will be used to implement functionality that will allow taxa unidentified at the genus level to be represented by their most specific known classification. *Disclaimer here about how this process will likely not be identical for every set of data but rather it is important to learn the general concepts taught here*  -->

<!-- ### Generating an Abundance plot with Phyloseq  -->

<!-- ### Generating an Abundance plot "Manually"  -->

<!-- Ensure that "tidyverse" , "reshape2", and "zoo" are installed and load them for use.  -->
<!-- ```{r, message=FALSE, warning=FALSE, results=FALSE} -->
<!-- library(tidyverse) -->
<!-- library(reshape2) -->
<!-- library(zoo) -->
<!-- ``` -->

<!-- In the event that the `OTU.table.with.taxalevel` data frame is needed elsewhere generate an identical table. For simplicity, the taxonomic classifications in one column will be used. Taxonomy separated by column can be removed for ease of performance in the following steps and be added back in later.  -->
<!-- ```{r} -->
<!-- OTWT <- OTU.table.with.taxalevel -->
<!-- OTWT[,1:7] <- NULL -->
<!-- colnames(OTWT)[1] <- "TaxaName" #renames the column with consolidated taxonomy -->
<!-- ``` -->

<!-- In the next step the data will be melted into long format. This is to make each row a distinct taxa that is found within a sample. From this there may be duplicates of the same taxa by row if it is found in more than one sample. Before this is done ensure the column names are set as characters. From this a the generated data frame will list taxa by row and their corresponding sample and amount in said sample by column. Lastly, remove taxa where that do not appear in a sample.  -->
<!-- ```{r} -->
<!-- char.cols <- as.character(colnames(OTWT)[2:97]) # ensures col names are characters and makes the measure.vars easy to input -->
<!-- melt.otu <- melt(OTWT, id.vars = c("TaxaName"), -->
<!--             measure.vars = char.cols, -->
<!--             variable.name = "SampleID", value.name = "count") -->
<!-- melt.otu.filt <- filter(melt.otu, count > 0) -->
<!-- ```  -->

<!-- Now we must merge the melted OTU table with the metadata. To do this we both data frames must have a shared column by which to merge. To do so, create a second metadata table with a "SampleID" column. Merge the melted OTU table with the metadata. Ensure that none of the column names conflict in newly merged dataframe. In this case "Family" the taxonomic level, and "Family" the sample grouping conflict - as such the sample grouping will be changed to "FamilyNum".  -->
<!-- ```{r} -->
<!-- metadata2 <- rownames_to_column(metadata, var = "SampleID") -->
<!-- meta.tax <- merge(melt.otu.filt, metadata2) -->
<!-- colnames(meta.tax)[7] <- "FamilyNum" -->
<!-- ``` -->

<!-- Now the taxa levels which were condensed into one column can be restored to their separate columns for ease of readability, as a new data frame.    -->
<!-- ```{r} -->
<!-- meta.tax.na <- ldply(str_split(string = meta.tax$TaxaName, pattern=";"), rbind) -->
<!-- names(meta.tax.na) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species") -->
<!-- meta.tax2 <- as.data.frame(lapply(meta.tax.na, gsub, pattern=" ", replacement="")) -->
<!-- ``` -->

<!-- It may be desirable to represent some taxa at a higher level of classification (e.g. Genera instead of Phyla) however some species may be unidentified at this level. One option that may fix this is by allowing the highest known level of classification to be presented. To do this, the highest known classification level must apply to the level being examined.  -->
<!-- ```{r} -->
<!-- meta.tax2 <- data.frame(lapply(meta.tax2, function(x) {gsub("[a-z]__", "", x)})) # removes the "^__" before taxa  -->
<!-- meta.tax2 <- meta.tax2 %>% mutate_all(na_if,"") # makes blanks go to NAs -->
<!-- meta.tax2 <- data.frame(t(apply(meta.tax2,1,function(x) na.locf(x)))) #replaces NAs in taxa with last known level of classification -->
<!-- ``` -->

<!-- Now we can bind this back to `meta.tax` and remove the condensed taxa column.  -->
<!-- ```{r} -->
<!-- meta.tax<- cbind(meta.tax2[,1:7],meta.tax) -->
<!-- meta.tax$TaxaName <- NULL -->
<!-- ``` -->

<!-- Now the most prevalent taxa must be found, in this case the top 15. Before doing so, ensure that the counts of taxa in each sample is set as numeric. However in order to perform this task we must first unload the plyr and reshape2(dependent on plyr) packages since they conflict with tidyverse.  -->
<!-- ```{r} -->
<!-- detach("package:reshape2", unload = TRUE) -->
<!-- detach("package:plyr", unload = TRUE) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- meta.tax$count <- as.numeric(meta.tax$count) #don't need to run if count col is already numeric -->
<!-- top.bacteria <- meta.tax %>% -->
<!--   group_by(., Genus) %>% -->
<!--   summarise(., total_reads = sum(count)) %>% -->
<!--   arrange(desc(total_reads)) %>% -->
<!--   slice(1:15) %>% pull(., Genus) -->
<!-- ``` -->

<!-- Now create a column that indicates whether each row corresponds to a top taxa or is of low abundance as designated by `top.bacteria` -->
<!-- ```{r} -->
<!-- meta.tax$taxa <- ifelse(meta.tax$Genus %in% top.bacteria, as.character(meta.tax$Genus), "low abundant") -->
<!-- ``` -->

<!-- Lastly, generate a multi-panel bar-plot organized by day and individual to display the top genera (or highest known classification) in ggplot2.  -->
<!-- ```{r} -->
<!-- meta.tax %>% -->
<!--   group_by(., Individual, Day, Genus, FamilyNum) %>% -->
<!--   summarise(., readsum = sum(count)) %>% -->
<!--   mutate(., taxa = ifelse(readsum > 200, as.character(Genus), "low abundant")) %>% -->
<!--   ggplot() + aes(x = Day, weight = readsum, fill = taxa)+ -->
<!--   geom_bar(position = position_fill()) + -->
<!--   scale_y_continuous(labels = scales::percent_format()) + -->
<!--   facet_wrap(. ~ FamilyNum + Individual, scales = "free") + -->
<!--   theme_dark() -->
<!-- ``` -->









